### PR TITLE
How to setup a virtual environment in Scala?

### PR DESCRIPTION
**Pull request by mariogalic**
To help with onboarding Python developers into Scala ecosystem.

### CODE REVIEW

**Review by Jasper-M — COMMENTED** [review]
I think you're getting a bit caught up in explaining python. I would leave that off, it's not a python tutorial. I'm also not sure I would say that with sbt dependencies are "installed". The fact that ivy or coursier keep local caches is pretty much orthogonal to how dependency management works, or how it is different from the python situation.

**Comment by mariogalic** [comment]
Thanks @Jasper-M for feedback! I am not sure how to proceed without being able to compare it to Python. I tried to emulate side-by-side comparison approach taken in the rest of "Scala for Python Developers" page. Shall I rename "installed" to "downloaded"? IMO the most visible differences for newcomer dependencies are not duplicated across projects and environmental variables are not modified. Why is it orthogonal?

**Comment by Jasper-M** [comment]
> I am not sure how to proceed without being able to compare it to Python.

I would start with a short description of how python and python modules are by default installed globally for the whole system, and how virtual environments are a solution to that. Then how Scala fundamentally doesn't have that issue because of how build tools and dependency management for JVM works. That's also a bit what I meant with the orthogonal comment. But I think you're probably right that it's definitely worth mentioning in this context as it's a big difference from python.

But that's all my opinion. Someone else may feel differently.

**Comment by LPTK** [comment]
> Shall I rename "installed" to "downloaded"? IMO the most visible differences for newcomer are that dependencies are not duplicated across projects and environmental variables are not modified. Why is it orthogonal?

It's completely different. You can `rm -rf` SBT's cache, and the only difference you'll experience is that first-time build times after that will take longer as the artifacts will need redownloading. So SBT's cache is an implementation detail, which does not affect the semantics of the build (hence "orthogonal").

Compare that to _installing_ packages, which destructively modifies the global (or virtual) state of the build tool. Imagine what would happen if you started deleting files you have installed as part of a Python package.

I think this difference is fundamental and super important, even for beginners.

File: _overviews/scala3-book/scala-for-python-devs.md [hunk 1]
```diff
@@ -1291,6 +1291,29 @@ Follow the links below for more details:
 - Infix methods
 - Macros and metaprogramming
 
+## How to setup a virtual environnement in Scala?
+
```
Comment (martijnhoekstra) [inline _overviews/scala3-book/scala-for-python-devs.md:1295]: I'd love if there were a "most important takeaway" right on top before you get to the introduction about what python does.

In scala, there is no need for virtual environments.

Then after that comes a description of the python situation and how building with sbt differs from that. A reader could stop reading after that first sentence, or read more if they want to convince themselves it's true.

**Comment by SethTisue** [comment]
@LPTK re-review?

**Review by LPTK — CHANGES_REQUESTED** [review]
Ok, I read through it and found that the writing could be improved in many places. Please apply my patches, as there are many minor grammatical mistakes in the current version.

**Comment by mariogalic** [comment]
I've applied the patches. Please feel free to make any direct changes. As a side note build is failing because GraalVM site seems to have removed Scala link so it results in 404.

**Comment by LPTK** [comment]
I don't want to be a grammar nazi, but the writing still feels broken in many places. Just looking at the first sentence:

> In Scala there is no need to explicitly setup an equivalent of Python's virtual environment.

"setup" is not a verb; "set up" is. And "Python" does not have one virtual environment, so "Python's virtual environment" is the wrong formulation. That's why I proposed "set up the equivalent of a Python virtual environment".

I'm going over the rest, but you get the idea. I guess it's not that important though, so feel free to merge.

**Comment by mariogalic** [comment]
I must have messed up interactive rebase because I thought I applied all your patches. I will fix it.

**Comment by mariogalic** [comment]
@LPTK Your patches seem to be applied correctly. Maybe it was showing outdated github page for you? For example first sentence does have "set up" instead of "setup".

**Comment by LPTK** [comment]
@mariogalic ok cool, I guess I was reading the wrong diff, thanks 👍

**Review by LPTK — APPROVED** [review]

**Comment by mariogalic** [comment]
Any ideas for improving the title? Perhaps

- Virtual Environments for Scala
- Creating virtual environments for Scala
- Set up virtual environments for Scala
- Scala virtual environments

**Comment by LPTK** [comment]
@mariogalic "Scala and Virtual Environments"?

**Comment by SethTisue** [comment]
is this ready for merge?

**Comment by mariogalic** [comment]
@SethTisue Is the question for me? I think I addressed all the comments. Do let me know if anything else is needed.

**Comment by mariogalic** [comment]
Anyone interested in merging this? I do not have write access.

**Comment by LPTK** [comment]
Me neither.

**Pull request merged**